### PR TITLE
HDDS-9015. Block CSR request in SCM for "hdds.x509.rootca.certificate.polling.interval" time period

### PR DESCRIPTION
**Pull request by ChenSammi**
https://issues.apache.org/jira/browse/HDDS-9015

Tested it manually

```
bash-4.2$ for i in 1..100 ; do ozone admin scm transfer -n b49649d7-7aa1-4b6e-bb64-73bd9dd1acb6; done
com.google.protobuf.ServiceException: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdds.scm.exceptions.SCMException): This action is prohibited for PT10S due to root CA and sub CA rotation have just finished. Please try the operation later again.
	at org.apache.hadoop.hdds.scm.server.SCMClientProtocolServer.transferLeadership(SCMClientProtocolServer.java:815)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.transferScmLeadership(StorageContainerLocationProtocolServerSideTranslatorPB.java:1219)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.processRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:685)
	at org.apache.hadoop.hdds.server.OzoneProtocolMessageDispatcher.processRequest(OzoneProtocolMessageDispatcher.java:87)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.submitRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:221)
	at org.apache.hadoop.hdds.protocol.proto.StorageContainerLocationProtocolProtos$StorageContainerLocationProtocolService$2.callBlockingMethod(StorageContainerLocationProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.processCall(ProtobufRpcEngine.java:484)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:595)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:573)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1227)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1094)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1017)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1899)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:3048)
, while invoking $Proxy20.submitRequest over nodeId=scmNodeId,nodeAddress=scm/172.20.0.3:9860 after 1 failover attempts. Trying to failover after sleeping for 1000ms. Current retry count: 1.
com.google.protobuf.ServiceException: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdds.scm.exceptions.SCMException): This action is prohibited for PT10S due to root CA and sub CA rotation have just finished. Please try the operation later again.
	at org.apache.hadoop.hdds.scm.server.SCMClientProtocolServer.transferLeadership(SCMClientProtocolServer.java:815)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.transferScmLeadership(StorageContainerLocationProtocolServerSideTranslatorPB.java:1219)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.processRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:685)
	at org.apache.hadoop.hdds.server.OzoneProtocolMessageDispatcher.processRequest(OzoneProtocolMessageDispatcher.java:87)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.submitRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:221)
	at org.apache.hadoop.hdds.protocol.proto.StorageContainerLocationProtocolProtos$StorageContainerLocationProtocolService$2.callBlockingMethod(StorageContainerLocationProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.processCall(ProtobufRpcEngine.java:484)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:595)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:573)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1227)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1094)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1017)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1899)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:3048)
, while invoking $Proxy20.submitRequest over nodeId=scmNodeId,nodeAddress=scm/172.20.0.3:9860 after 2 failover attempts. Trying to failover after sleeping for 1000ms. Current retry count: 2.
com.google.protobuf.ServiceException: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdds.scm.exceptions.SCMException): This action is prohibited for PT10S due to root CA and sub CA rotation have just finished. Please try the operation later again.
	at org.apache.hadoop.hdds.scm.server.SCMClientProtocolServer.transferLeadership(SCMClientProtocolServer.java:815)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.transferScmLeadership(StorageContainerLocationProtocolServerSideTranslatorPB.java:1219)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.processRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:685)
	at org.apache.hadoop.hdds.server.OzoneProtocolMessageDispatcher.processRequest(OzoneProtocolMessageDispatcher.java:87)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.submitRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:221)
	at org.apache.hadoop.hdds.protocol.proto.StorageContainerLocationProtocolProtos$StorageContainerLocationProtocolService$2.callBlockingMethod(StorageContainerLocationProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.processCall(ProtobufRpcEngine.java:484)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:595)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:573)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1227)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1094)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1017)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1899)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:3048)
, while invoking $Proxy20.submitRequest over nodeId=scmNodeId,nodeAddress=scm/172.20.0.3:9860 after 3 failover attempts. Trying to failover after sleeping for 1000ms. Current retry count: 3.
com.google.protobuf.ServiceException: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdds.scm.exceptions.SCMException): This action is prohibited for PT10S due to root CA and sub CA rotation have just finished. Please try the operation later again.
	at org.apache.hadoop.hdds.scm.server.SCMClientProtocolServer.transferLeadership(SCMClientProtocolServer.java:815)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.transferScmLeadership(StorageContainerLocationProtocolServerSideTranslatorPB.java:1219)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.processRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:685)
	at org.apache.hadoop.hdds.server.OzoneProtocolMessageDispatcher.processRequest(OzoneProtocolMessageDispatcher.java:87)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.submitRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:221)
	at org.apache.hadoop.hdds.protocol.proto.StorageContainerLocationProtocolProtos$StorageContainerLocationProtocolService$2.callBlockingMethod(StorageContainerLocationProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.processCall(ProtobufRpcEngine.java:484)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:595)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:573)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1227)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1094)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1017)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1899)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:3048)
, while invoking $Proxy20.submitRequest over nodeId=scmNodeId,nodeAddress=scm/172.20.0.3:9860 after 4 failover attempts. Trying to failover after sleeping for 1000ms. Current retry count: 4.
com.google.protobuf.ServiceException: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdds.scm.exceptions.SCMException): This action is prohibited for PT10S due to root CA and sub CA rotation have just finished. Please try the operation later again.
	at org.apache.hadoop.hdds.scm.server.SCMClientProtocolServer.transferLeadership(SCMClientProtocolServer.java:815)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.transferScmLeadership(StorageContainerLocationProtocolServerSideTranslatorPB.java:1219)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.processRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:685)
	at org.apache.hadoop.hdds.server.OzoneProtocolMessageDispatcher.processRequest(OzoneProtocolMessageDispatcher.java:87)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.submitRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:221)
	at org.apache.hadoop.hdds.protocol.proto.StorageContainerLocationProtocolProtos$StorageContainerLocationProtocolService$2.callBlockingMethod(StorageContainerLocationProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.processCall(ProtobufRpcEngine.java:484)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:595)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:573)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1227)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1094)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1017)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1899)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:3048)
, while invoking $Proxy20.submitRequest over nodeId=scmNodeId,nodeAddress=scm/172.20.0.3:9860 after 5 failover attempts. Trying to failover after sleeping for 1000ms. Current retry count: 5.
com.google.protobuf.ServiceException: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdds.scm.exceptions.SCMException): This action is prohibited for PT10S due to root CA and sub CA rotation have just finished. Please try the operation later again.
	at org.apache.hadoop.hdds.scm.server.SCMClientProtocolServer.transferLeadership(SCMClientProtocolServer.java:815)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.transferScmLeadership(StorageContainerLocationProtocolServerSideTranslatorPB.java:1219)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.processRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:685)
	at org.apache.hadoop.hdds.server.OzoneProtocolMessageDispatcher.processRequest(OzoneProtocolMessageDispatcher.java:87)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.submitRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:221)
	at org.apache.hadoop.hdds.protocol.proto.StorageContainerLocationProtocolProtos$StorageContainerLocationProtocolService$2.callBlockingMethod(StorageContainerLocationProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.processCall(ProtobufRpcEngine.java:484)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:595)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:573)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1227)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1094)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1017)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1899)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:3048)
, while invoking $Proxy20.submitRequest over nodeId=scmNodeId,nodeAddress=scm/172.20.0.3:9860 after 6 failover attempts. Trying to failover after sleeping for 1000ms. Current retry count: 6.
com.google.protobuf.ServiceException: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdds.scm.exceptions.SCMException): This action is prohibited for PT10S due to root CA and sub CA rotation have just finished. Please try the operation later again.
	at org.apache.hadoop.hdds.scm.server.SCMClientProtocolServer.transferLeadership(SCMClientProtocolServer.java:815)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.transferScmLeadership(StorageContainerLocationProtocolServerSideTranslatorPB.java:1219)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.processRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:685)
	at org.apache.hadoop.hdds.server.OzoneProtocolMessageDispatcher.processRequest(OzoneProtocolMessageDispatcher.java:87)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.submitRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:221)
	at org.apache.hadoop.hdds.protocol.proto.StorageContainerLocationProtocolProtos$StorageContainerLocationProtocolService$2.callBlockingMethod(StorageContainerLocationProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.processCall(ProtobufRpcEngine.java:484)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:595)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:573)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1227)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1094)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1017)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1899)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:3048)
, while invoking $Proxy20.submitRequest over nodeId=scmNodeId,nodeAddress=scm/172.20.0.3:9860 after 7 failover attempts. Trying to failover after sleeping for 1000ms. Current retry count: 7.
com.google.protobuf.ServiceException: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdds.scm.exceptions.SCMException): This action is prohibited for PT10S due to root CA and sub CA rotation have just finished. Please try the operation later again.
	at org.apache.hadoop.hdds.scm.server.SCMClientProtocolServer.transferLeadership(SCMClientProtocolServer.java:815)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.transferScmLeadership(StorageContainerLocationProtocolServerSideTranslatorPB.java:1219)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.processRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:685)
	at org.apache.hadoop.hdds.server.OzoneProtocolMessageDispatcher.processRequest(OzoneProtocolMessageDispatcher.java:87)
	at org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocolServerSideTranslatorPB.submitRequest(StorageContainerLocationProtocolServerSideTranslatorPB.java:221)
	at org.apache.hadoop.hdds.protocol.proto.StorageContainerLocationProtocolProtos$StorageContainerLocationProtocolService$2.callBlockingMethod(StorageContainerLocationProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.processCall(ProtobufRpcEngine.java:484)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:595)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:573)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1227)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1094)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1017)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1899)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:3048)
, while invoking $Proxy20.submitRequest over nodeId=scmNodeId,nodeAddress=scm/172.20.0.3:9860 after 8 failover attempts. Trying to failover after sleeping for 1000ms. Current retry count: 8.
Transfer leadership successfully to b49649d7-7aa1-4b6e-bb64-73bd9dd1acb6.
```

### CODE REVIEW

**Review by fapifta — COMMENTED** [review]
Thank you for your work @ChenSammi, overall the changes look good, I have a general quesiton though:
As I understand, during the rootCA rotation we start a task that will cancel the post rotation phase if leadership is transferred to an other SCM, this is done in the notifyStatusChanged method within the RootCARotationManager.
Other SCMs during the prepare ack will start a similar timer, so that once  a leader change happens, those timers remain active, again if I understand correctly.

However I am not sure what will happen if there are two subsequent leadership transfer like: SCM1->SCM2-SCM1? But my understanding is that in this case the notifyStatusChanged would be a place to check if we are still in the post processing period.

There is one more case, when all 3 SCMs are restarted at once during the processing period, in which case the SCM does not seem to activate the post processing period and does not disable certificate signing further, which can be a problem if other services are not restarted together with the SCMs and their poller does not run.

**Comment by ChenSammi** [comment]
> Thank you for your work @ChenSammi, overall the changes look good, I have some general quesiton though: As I understand, during the rootCA rotation we start a task that will cancel the post rotation phase if leadership is transferred to an other SCM, this is done in the notifyStatusChanged method within the RootCARotationManager. Other SCMs during the prepare ack will start a similar timer, so that once a leader change happens, those timers remain active, again if I understand correctly.
> 
> However I am not sure what will happen if there are two subsequent leadership transfer like: SCM1->SCM2-SCM1? But my understanding is that in this case the notifyStatusChanged would be a place to check if we are still in the post processing period.
> 
> There is one more case, when all 3 SCMs are restarted at once during the processing period, in which case the SCM does not seem to activate the post processing period and does not disable certificate signing further, which can be a problem if other services are not restarted together with the SCMs and their poller does not run.

@fapifta , good catch. A new commit is uploaded to handle the leader transfer and service restart cases.

**Comment by ChenSammi** [comment]
The failed unit test is caused by "no space left on device".

**Review by fapifta — COMMENTED** [review]
Hi @ChenSammi,
thank you for addressing my comments, and sorry for the delay, I was occupied last week with a good couple of things.

The patch looks good, I noticed three minor things, actually the findbugs one is probably the most important to do not introduce new warnings with out of order checks of related merges.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/RootCARotationManager.java [hunk 1]
```diff
@@ -710,4 +761,52 @@ public boolean shouldSkipRootCert(String newRootCertId) throws IOException {
...
+      // this should not happen
+      throw new RuntimeException("Root CA certificate ID " +
+          cert.getSerialNumber() + " in stateful storage is bigger than " +
+          "current scm's root CA certificate ID " + rootCert.getSerialNumber());
+    }
+
+    Date issueTime = rootCert.getNotBefore();
+    Date now = Calendar.getInstance().getTime();
+    Duration gap = Duration.between(issueTime.toInstant(), now.toInstant());
+    gap = gap.minus(rootCertPollInterval);
```
[the remark this answers is not in the document]
Comment (ChenSammi) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/RootCARotationManager.java:801]: rootCertPollInterval is a final Duration used as a constant variable in the class. I would prefer not to mutate it.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java [hunk 2]
```diff
@@ -208,16 +206,31 @@ public String getCertificate(
...
+    if (rotationManager != null) {
+      if (rotationManager.isRotationInProgress() && !isScmCertRenew) {
+        throw new SCMException("Root CA and Sub CA rotation is in-progress." +
+            " Please try the operation later again.",
+            ResultCodes.CA_ROTATION_IN_PROGRESS);
+      }
+      if (rotationManager.isPostRotationInProgress()) {
+        SecurityConfig securityConfig = new SecurityConfig(config);
+        throw new SCMException("This action is prohibited for " +
+            securityConfig.getRootCaCertificatePollingInterval() +
```
[the remark this answers is not in the document]
Comment (ChenSammi) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java:226]: How about this wording? 

"This action is prohibited due to that root CA and sub CA rotation have just finished." 
"The prohibition state will last at most " +  securityConfig.getRootCaCertificatePollingInterval().
"Please try the operation later again."

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/RootCARotationManager.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (fapifta) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/RootCARotationManager.java:801]: I see, I am fine with this approach, thank you for clarifying.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (fapifta) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java:226]: I like this wording more, if you would change it that would be good. Thank you!

[the remark this answers is not in the document]
Comment (ChenSammi) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java]: Thank you @fapifta for reviewing the code.

**Review by fapifta — COMMENTED** [review]
Hi Sammi, thank you very much for addressing the comments.

I just realized that the SCMClientProtocolServer, the SCMSecurityProtocolServer, and StorageContainerManager all share this code for the rotation and post rotation state check, and in all cases, the logic is that the code asks the rootCARotationManager about the state, and if the state is x, it throws an exception, and now one of the places have a different exception message as of now.

I suggest to push these checks along with the throwing the exception to one place? I think about a method on the rootCARotationManager, that asserts whether we are in a rotation or in the wait period, and throws or does not throw the exception accordingly?
Also I thought a bit more on the wording, what do you think about having something like this:
"A new root CA certificate is being applied to the cluster, during this period <operation_name> operation is prohibited. Applying the root CA certificate takes <polling interval> minutes.". To this message the operation name can arrive to the checking method via a parameter.

**Comment by ChenSammi** [comment]
Make sense, addressed @fapifta .

**Review by fapifta — APPROVED** [review]
Thank you Sammi!
